### PR TITLE
Fixes missing prompt template or system template

### LINKED ISSUE #2402: [BUG]If only system_template is provided and prompt_template is missing,error occurred!
### Description

The CrewAI framework throws an AttributeError: 'NoneType' object has no attribute 'replace' when attempting to execute a Crew with an Agent configured with only the system_template parameter, while the prompt_template is left undefined (i.e., None).
And ,if prompt_template proveded also,but response_template not provided,error occurred also!

### Steps to Reproduce

simple example 

### Expected behavior

no error,if these three parameters are optional

### Screenshots/Code snippets

from crewai import Agent, Task, Crew

# Incorrect Agent configuration (missing prompt_template)
designer = Agent(
  name="WeChat Layout Expert",
  role="WeChat Layout Expert",
  goal="Design and optimize the generated article for WeChat public accounts.",
  backstory="You are a professional WeChat public account layout designer, familiar with WeChat layout rules.",
  allow_delegation=False,
  memory=True,
  max_rpm=150,
  system_template="You are a professional WeChat public account layout designer...", # system_template present
  # prompt_template is missing
)

# Task and Crew setup (example)
task = Task(description="Layout task", agent=designer)
crew = Crew(agents=[designer], tasks=[task])

# Attempting to execute the Crew will result in the error
crew.kickoff()

### Operating System

Windows 11

### Python Version

3.10

### crewAI Version

102

### crewAI Tools Version

no tool

### Virtual Environment

Venv

### Evidence

Traceback (most recent call last):
  File "C:\Users\X\Desktop\test\src\test\main.py", line 26, in run
    Test().crew().kickoff(inputs=inputs)
    ^^^^^^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 36, in __init__
    self.map_all_task_variables()
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 203, in map_all_task_variables
    self._map_task_variables(
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 236, in _map_task_variables
    self.tasks_config[task_name]["agent"] = agents[agent_name]()
                                            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\utils.py", line 11, in memoized_func
    cache[key] = func(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\X\Desktop\test\src\test\crew.py", line 37, in designer
    return Agent(
           ^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\pydantic\main.py", line 214, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 133, in post_init_setup
    self._setup_agent_executor()
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 143, in _setup_agent_executor
    self.set_cache_handler(self.cache_handler)
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agents\agent_builder\base_agent.py", line 341, in set_cache_handler
    self.create_agent_executor()
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 291, in create_agent_executor
    ).task_execution()
      ^^^^^^^^^^^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\utilities\prompts.py", line 41, in task_execution
    "prompt": self._build_prompt(
              ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\X\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\utilities\prompts.py", line 67, in _build_prompt
    prompt = prompt_template.replace(
             ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'replace'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\X\Desktop\test\src\test\main.py", line 175, in <module>
    test()
  File "C:\Users\X\Desktop\test\src\test\main.py", line 171, in test
    run(inputs)
  File "C:\Users\X\Desktop\test\src\test\main.py", line 28, in run
    raise Exception(f"An error occurred while running the crew: {e}")
Exception: An error occurred while running the crew: 'NoneType' object has no attribute 'replace'
---------------
Traceback (most recent call last):
  File "C:\Users\XX\Desktop\test\src\test\main.py", line 26, in run
    Test().crew().kickoff(inputs=inputs)
    ^^^^^^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 36, in __init__
    self.map_all_task_variables()
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 203, in map_all_task_variables
    self._map_task_variables(
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\crew_base.py", line 236, in _map_task_variables
    self.tasks_config[task_name]["agent"] = agents[agent_name]()
                                            ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\project\utils.py", line 11, in memoized_func
    cache[key] = func(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XX\Desktop\test\src\test\crew.py", line 37, in designer
    return Agent(
           ^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\pydantic\main.py", line 214, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 133, in post_init_setup
    self._setup_agent_executor()
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 143, in _setup_agent_executor
    self.set_cache_handler(self.cache_handler)
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agents\agent_builder\base_agent.py", line 341, in set_cache_handler
    self.create_agent_executor()
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\agent.py", line 291, in create_agent_executor
    ).task_execution()
      ^^^^^^^^^^^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\utilities\prompts.py", line 41, in task_execution
    "prompt": self._build_prompt(
              ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XX\AppData\Local\Programs\Python\Python311\Lib\site-packages\crewai\utilities\prompts.py", line 70, in _build_prompt
    response = response_template.split("{{ .Response }}")[0]
               ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'split'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\XX\Desktop\test\src\test\main.py", line 175, in <module>
    test()
  File "C:\Users\XX\Desktop\test\src\test\main.py", line 171, in test
    run(inputs)
  File "C:\Users\XX\Desktop\test\src\test\main.py", line 28, in run
    raise Exception(f"An error occurred while running the crew: {e}")
Exception: An error occurred while running the crew: 'NoneType' object has no attribute 'split'

### Possible Solution

i dont know 

### Additional context

use system_template（prompt_template、response_template provided）,gemini cant run correctly,,,but deepseek 、qwq32b run no error

### PR DESCRIPTION
**Pull request by Vidit-Ostwal**
Fixes #2402

Built on top of #2403, 
Changed the structure of files and added test cases in the test_agent itself.

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for PR #2408

## Overview
This PR addresses issue #2402, enhancing the handling of templates within the crewAI framework. The primary goal is to bolster the system's resilience when encountering missing templates.

### 1. src/crewai/utilities/prompts.py
#### Positive Changes
- The error handling for missing templates has been significantly improved.
- There is now a fallback mechanism to a default format which enhances user experience.
- The separation of template handling logic has been improved, promoting better readability.

#### Issues Identified
1. **Redundant Conditions**: 
   - The condition `if not system_template or not prompt_template` could lead to unexpected behavior. A more nuanced condition could improve clarity.
   
   **Suggestion**: Implement a more detailed check for critical vs. optional templates to tighten the logic and prevent potential edge cases.

2. **Missing Input Validation**:
   - The current implementation lacks validation for the format of the templates, which can allow erroneous types to be accepted without feedback.

3. **Incomplete Error Handling**:
   - The current implementation does not provide clear error messages for cases involving malformed templates, which can hinder debugging.

#### Suggested Code Improvement

```python
def _build_prompt(
        self,
        components,
        system_template=None,
        prompt_template=None,
        response_template=None,
    ) -> str:
        # Input validations should be added here.
        # Example code for validation
        if system_template and not isinstance(system_template, str):
            raise ValueError("system_template must be a string")
```

### 2. Tests
#### Positive Aspects
- Good test coverage for new functionalities and different template scenarios.
- Clear assertions to validate expected behavior of the system.

#### Issues Identified
1. **Duplicate Test Cases**: 
   - Similar tests are found across multiple files, which may lead to maintenance challenges.
   
2. **Inconsistent Test Organization**: 
   - Tests should be organized logically to improve clarity and reduce redundant code.

3. **Missing Edge Cases**: 
   - Some potential template combinations were not adequately tested, which could introduce unforeseen issues.

#### Suggested Test Improvement

```python
@pytest.mark.parametrize("templates", [
    {"system_template": "Example Template"},
    {"prompt_template": "Prompt Template"},
])
def test_agent_template_combinations(templates):
    # Test various template combinations with clear assertions.
```

### General Recommendations
1. **Code Organization**:
   - Consolidate all template-related tests into a dedicated file for better maintainability.
   - Move utility functions into a dedicated module if necessary.

2. **Error Handling**:
   - Improve logging mechanisms for template processing errors for enhanced traceability.
   - Consider introducing custom exceptions to clarify errors related to template handling.

3. **Testing**:
   - Integrate property-based tests and edge case scenarios to fully validate template handling logic.
   - Consider adding integration tests that utilize real-world template examples for comprehensive test coverage.

4. **Documentation**:
   - Enhance documentation regarding valid and invalid template formats, including the fallback behavior of the system.
   - Include warning notes about the necessity for input validation.

### Impact Analysis
- **Positive**: The changes enhance the reliability and robustness of template handling.
- **Neutral**: Minor performance overhead from additional checks, though generally acceptable.
- **Risk**: Potential issues with backward compatibility; thorough testing recommended to ensure existing functionality remains unaffected.

The changes made in this PR are a step in the right direction towards improving the resilience of template handling within crewAI. Adhering to the suggested improvements will further solidify the implementation and mitigate risks in future enhancements.

**Review by lucasgomide — APPROVED** [review]

**Comment by lucasgomide** [comment]
@Vidit-Ostwal do you mind to sync your branch?

File: src/crewai/utilities/prompts.py [hunk 1]
```diff
@@ -54,10 +54,12 @@ def _build_prompt(
         response_template=None,
     ) -> str:
         """Constructs a prompt string from specified components."""
-        if not system_template and not prompt_template:
+        if not system_template or not prompt_template:
```
Comment (lucasgomide) [inline src/crewai/utilities/prompts.py:57]: Would be great enhance some documentation to keep explict that they are dependents on each other

File: docs/concepts/agents.mdx [hunk 2]
```diff
@@ -255,7 +255,11 @@ custom_agent = Agent(
 - `response_template`: Formats agent responses
 
 <Note>
-When using custom templates, you can use variables like `{role}`, `{goal}`, and `{input}` in your templates. These will be automatically populated during execution.
+When using custom templates, ensure that both `system_template` and `prompt_template` are defined. The `response_template` is optional but recommended for consistent output formatting.
+</Note>  
+
+<Note>
+When using custom templates, you can use variables like `{role}`, `{goal}`, and `{backstory}` in your templates. These will be automatically populated during execution.
```
Comment (Vidit-Ostwal) [inline docs/concepts/agents.mdx:262]: Have changed input to backstory, following Line 82 in the prompts.py file

**Review by lucasgomide — APPROVED** [review]

**Comment by mplachta** [comment]
Disclaimer: This review was made by a crew of AI Agents.

Summary of Key Findings
-----------------------
This PR robustly addresses issue #2402 by enhancing the prompt-building logic to gracefully handle missing template components (`system_template`, `prompt_template`, and optionally `response_template`) when constructing prompts for agents. The patch updates the `_build_prompt` function to fallback sensibly if required templates are missing and avoid runtime errors related to missing template segments. New comprehensive tests were introduced to validate these scenarios but were later consolidated into the main tests/agent_test.py suite to maintain a well-organized test structure. Additionally, documentation was improved to clearly communicate template requirements and usage for users defining custom agents.

Specific Code Improvements with Examples
----------------------------------------

1. **Improved Fallback Logic in `_build_prompt` (src/crewai/utilities/prompts.py):**

   - The method now explicitly checks if either `system_template` or `prompt_template` is missing, and falls back to joining component slices rather than raising exceptions.
   - Handles missing `response_template` by excluding it gracefully from the constructed prompt.
   - Replacements for template variables `{goal}`, `{backstory}`, and `{role}` in the prompt are chained clearly to ensure all placeholders are replaced.

   Example snippet after refactor:

   ```python
   def _build_prompt(
       self, components, system_template=None, prompt_template=None, response_template=None
   ) -> str:
       """Constructs a prompt string from specified components."""
       if not system_template or not prompt_template:
           # Fallback to default component concatenation if templates missing
           prompt_parts = [self.i18n.slice(component) for component in components]
           prompt = "".join(prompt_parts)
       else:
           system = system_template
           prompt = prompt_template.replace("", "".join(self.i18n.slice("task")))
           if response_template:
               response = response_template.split("")[0]
               prompt = f"{system}\n{prompt}\n{response}"
           else:
               prompt = f"{system}\n{prompt}"

       prompt = (
           prompt.replace("{goal}", str(self.agent.goal))
                 .replace("{backstory}", str(self.agent.backstory))
                 .replace("{role}", str(self.agent.role))
       )
       return prompt
   ```

   **Suggestions:**
   - Use explicit calls to `str()` when doing replacements to avoid issues if agent attributes are non-string.
   - Consider flattening conditionals to reduce nesting and improve readability.
   - Enhance docstring to describe fallback behavior in detail.

2. **Test Improvements (tests/agent_test.py):**

   - Added tests verifying agents can be instantiated with only `system_template`, only `prompt_template`, or missing `response_template`, without raising errors.
   - Included attribute assertions verifying the presence and correctness of key agent properties.
   
   Suggested further enhancement:

   ```python
   def test_agent_execution_with_missing_templates():
       agent = Agent(
           role="Test Role",
           goal="Test Goal",
           backstory="Test Backstory",
           system_template="System template content",
           # prompt_template deliberately missing
       )
       task = Task(description="Some task", agent=agent, expected_output="Expected output")
       
       # Assert no exceptions during task execution
       try:
           agent.execute_task(task)
       except Exception as e:
           pytest.fail(f"Execution failed with missing prompt_template: {e}")
   ```

   This would confirm not only agent creation but also successful prompt construction and execution.

3. **Documentation Updates (docs/concepts/agents.mdx):**

   - Clearly states that `system_template` and `prompt_template` are required when using custom templates.
   - Marks `response_template` as optional but recommended.
   - Highlights available template variables `{role}`, `{goal}`, and `{backstory}` for user reference.
   - Suggest enhancement: use bullet lists or code blocks for better readability and visual clarity.

4. **Code Style and Consistency:**

   - Tests now consistently assert agent attributes with proper capitalization matching initialized values.
   - Minor lint fixes applied (in `tests/llm_test.py`).
   - Imports reordered in test files for clarity.
   
Historical Context and Learnings from Related PRs
-------------------------------------------------

- Original attempts to fix the issue added new test files (`test_fix_2402.py` and `tests/test_templates.py`) which were subsequently deleted and their tests merged into `tests/agent_test.py` to centralize test maintenance.
- Earlier versions tested agent instantiation but not comprehensive task execution paths; current tests in `agent_test.py` are improved but could be further enhanced to test execution success fully.
- Documentation evolved to explicitly define template expectations minimizing user errors and confusion.
- The patch follows established patterns of adding robust fallback checks for user-provided templates and tightly integrating tests and docs to reinforce feature usage expectations.

Implications for Related Files
------------------------------

- The changes in `_build_prompt` impact how all agents build their prompts; ensuring backward compatibility and smooth error handling is critical.
- Tests in `tests/agent_test.py` serve as the canonical location for template-related test coverage, replacing earlier ad-hoc test scripts.
- Documentation updates align user understanding with new fallback logic, potentially reducing support queries.
- Minor lint fixes improve overall code hygiene.

Final Recommendations
---------------------

- **Merge the patch**, as it fixes a significant robustness issue concerning prompt template handling.
- **Enhance tests further** by adding assertions on actual prompt output and task execution success beyond attribute checks.
- **Refine docstrings** in `_build_prompt` to provide explicit fallback behavior and parameter explanations.
- Consider **standardizing capitalization** in test inputs and assertions across the entire test suite to prevent flaky tests.
- Format documentation notes using lists and code blocks for improved readability and developer adoption.

This PR is a well-structured and comprehensive fix for issue #2402 with solid improvements in code quality, test coverage, and documentation. No critical or security concerns are evident. Minor enhancements recommended above will further strengthen maintainability and usability.

---

Thank you for the opportunity to review this contribution. The improvements on prompt template handling and fallback logic significantly enhance stability in agent prompt processing. The integration and cleanup of tests and documentation ensure long-term maintainability and clear user guidance.

I look forward to seeing these refinements incorporated and the PR merged.

**Review by lucasgomide — APPROVED** [review]

**Comment by Sourav-Nandy-ai** [comment]
Hi Vidit,

Issue #2408 is a great example of a silent-crash versus silent-drift tradeoff. Before the fix, an Agent configured with only system_template set would crash with AttributeError: 'NoneType' object has no attribute 'replace' inside Prompts._build_prompt(). The and to or fix in this PR is exactly the guard that prevents this. I confirmed this live: crewai==0.117.1 crashes on Agent.create_agent_executor() with that exact exception, while crewai==0.118.0 succeeds and produces a real assembled prompt string.

I maintain agent-eval (agent_regress), an OSS harness for statistically comparing agent behavior across versions. Testing this PR exposed a real gap in the tooling. Comparing two installed versions of the same package required cross-venv orchestration that the tool did not previously support, since one interpreter only ever has one crewai version importable at a time.

The new subprocess_runner() now closes that gap by dispatching each call to a different venv's Python instance via a stdin and stdout JSON contract. This update is exactly what made this pre-and-post comparison possible to run cleanly. You can find the repository here: https://github.com/RudrenduPaul/agent-eval.

It is worth flagging this for anyone encountering similar behavior. Post-fix, if only one of system_template or prompt_template is set, the agent silently falls back to the default template rather than using the custom one. There is no crash, but the system_template is quietly ignored. This matches the PR's documented intent and is not a bug report on my end. It is simply worth knowing if you are trying to debug why a custom template is not applying.

Best,
Sourav
